src/cls/tabular/cls_tabular_utils.cc

```diff
@@ -596,6 +596,655 @@ int processSkyFb(
+                        break;
+                    }
+                    case SDT_DOUBLE: {
+                        TypedPredicate<double>* p = \
+                                dynamic_cast<TypedPredicate<double>*>(pb);
+                        double agg_val = p->Val();
+                        flexbldr->Add(agg_val);
+                        break;
+                    }
+                    default:  assert(UnsupportedAggDataType==0);
+                }
+            }
+        });
+        // finalize the row's projected data within our flexbuf
+        flexbldr->Finish();
+
+        // build the return ROW flatbuf that contains the flexbuf data
+        auto row_data = flatbldr.CreateVector(flexbldr->GetBuffer());
+        delete flexbldr;
+
+        // assume no nullbits in the agg results. ?
+        nullbits_vector nb(2,0);
+        auto nullbits = flatbldr.CreateVector(nb);
+        int RID = -1;  // agg recs only, since these are derived data
+        flatbuffers::Offset<Tables::Record> row_off = \
+            Tables::CreateRecord(flatbldr, RID, nullbits, row_data);
+
+        // Continue building the ROOT flatbuf's dead vector and rowOffsets vec
+        dead_rows.push_back(0);
+        offs.push_back(row_off);
+    } //if aggs
+
+    // now build the return ROOT flatbuf wrapper
+    std::string query_schema_str;
+    for (auto it = query_schema.begin(); it != query_schema.end(); ++it) {
+        query_schema_str.append(it->toString() + "\n");
+    }
+
+    auto return_data_schema = flatbldr.CreateString(query_schema_str);
+    auto db_schema_name = flatbldr.CreateString(root.db_schema_name);
+    auto table_name = flatbldr.CreateString(root.table_name);
+    auto delete_v = flatbldr.CreateVector(dead_rows);
+    auto rows_v = flatbldr.CreateVector(offs);
+
+    auto table = CreateTable(
+        flatbldr,
+        root.data_format_type,
+        root.skyhook_version,
+        root.data_structure_version,
+        root.data_schema_version,
+        return_data_schema,
+        db_schema_name,
+        table_name,
+        delete_v,
+        rows_v,
+        offs.size());
+
+    // NOTE: the fb may be incomplete/empty, but must finish() else internal
+    // fb lib assert finished() fails, hence we must always return a valid fb
+    // and catch any ret error code upstream
+    flatbldr.Finish(table);
+
+    return errcode;
+} //processSkyFb_fbu_rows
+
+int processSkyFb_fbu_cols(
+    ceph::bufferlist wrapped_bls,
+    flatbuffers::FlatBufferBuilder& flatbldr,
+    schema_vec& data_schema,
+    schema_vec& query_schema,
+    predicate_vec& preds,
+    const char* fb,
+    const size_t fb_size,
+    std::string& errmsg,
+    const std::vector<uint32_t>& row_nums)
+{
+    int errcode = 0;
+    delete_vector dead_rows;
+    std::vector<flatbuffers::Offset<Tables::Record>> offs;
+
+    // get row ids to process. should be uniform for all structures.
+    sky_root starter_root = getSkyRoot(fb, fb_size, SFT_FLATBUF_UNION_COL);
```

Comment:
Oh i see now, starter_root is the outer root, and root is the inner ones in the case where you have multiple cols in a single fb_meta.  You mentioned that.   Ok we should think about this later regarding how to layout multiple cols, would be good to get your thoughts if this approach to fbu cols should remain long term or possibly something else.  But yeah its fine, this is just a general comment so we don't forget later :)